File: frontend/src/edl.js
// Pure Edit Decision List (EDL) model — the single source of truth for an edit.
// An EDL is an ordered array of segments { id, sourceStart, sourceEnd } over the
// original video's timeline. The transcript, timeline, and preview are all pure
// projections of this array, so they can never desync. Every operation returns a
// NEW array (immutability) — callers keep undo history by snapshotting.

let _idCounter = 0;
export function newId() {
  _idCounter += 1;
  return `seg_${Date.now().toString(36)}_${_idCounter}`;
}

const MIN_SEG = 0.04; // ~1 frame at 25fps; minimum meaningful segment length

export function defaultEdl(duration) {
  return [{ id: newId(), sourceStart: 0, sourceEnd: duration }];
}

export function segmentDuration(seg) {
  return seg.sourceEnd - seg.sourceStart;
}

export function totalDuration(edl) {
  return edl.reduce((acc, s) => acc + segmentDuration(s), 0);
}

// virtual time (position in the edited timeline) -> source time (in original video)
export function virtualToSource(edl, vt) {
  let acc = 0;
  for (let i = 0; i < edl.length; i++) {
    const d = segmentDuration(edl[i]);
    if (vt < acc + d || (i === edl.length - 1 && vt <= acc + d + 1e-6)) {
      return { segIndex: i, source: edl[i].sourceStart + (vt - acc) };
    }
    acc += d;
  }
  return null;
}

// source time -> virtual time; null if that source time was cut out of the edit
export function sourceToVirtual(edl, source) {
  let acc = 0;
  for (const seg of edl) {
    if (source >= seg.sourceStart && source <= seg.sourceEnd) {
      return acc + (source - seg.sourceStart);
    }
    acc += segmentDuration(seg);
  }
  return null;
}

export function splitAtVirtual(edl, vt) {
  const map = virtualToSource(edl, vt);
  if (!map) return edl;
  const { segIndex, source } = map;
  const seg = edl[segIndex];
  if (source - seg.sourceStart < MIN_SEG || seg.sourceEnd - source < MIN_SEG) {
    return edl; // too close to an edge to split meaningfully
  }
  const left = { id: newId(), sourceStart: seg.sourceStart, sourceEnd: source };
  const right = { id: newId(), sourceStart: source, sourceEnd: seg.sourceEnd };
  return [...edl.slice(0, segIndex), left, right, ...edl.slice(segIndex + 1)];
}

export function trimSegment(edl, segId, edge, newSource) {
  return edl.map((s) => {
    if (s.id !== segId) return s;
    if (edge === "start") {
      const v = Math.min(newSource, s.sourceEnd - MIN_SEG);
      return { ...s, sourceStart: Math.max(0, v) };
    }
    const v = Math.max(newSource, s.sourceStart + MIN_SEG);
    return { ...s, sourceEnd: v };
  });
}

export function deleteSegment(edl, segId) {
  return edl.filter((s) => s.id !== segId);
}

export function reorderSegment(edl, fromIndex, toIndex) {
  if (fromIndex === toIndex) return edl;
  const copy = [...edl];
  const [moved] = copy.splice(fromIndex, 1);
  copy.splice(toIndex, 0, moved);
  return copy;
}

// Remove the source-time range [rangeStart, rangeEnd] from whichever segments it
// overlaps, splitting/trimming as needed. Used by transcript word/selection delete.
export function deleteSourceRange(edl, rangeStart, rangeEnd) {
  const out = [];
  for (const s of edl) {
    if (rangeEnd <= s.sourceStart || rangeStart >= s.sourceEnd) {
      out.push(s); // no overlap
      continue;
    }
    if (rangeStart > s.sourceStart) {
      out.push({ id: newId(), sourceStart: s.sourceStart, sourceEnd: rangeStart });
    }
    if (rangeEnd < s.sourceEnd) {
      out.push({ id: newId(), sourceStart: rangeEnd, sourceEnd: s.sourceEnd });
    }
  }
  return out;
}

// Transcript projection: walk segments in EDL order, emit words whose source
// timestamps fall inside each segment (>=minKeepRatio of the word retained),
// annotated with their virtual start/end. Guarantees transcript == timeline order.
export function projectWords(edl, words, minKeepRatio = 0.5) {
  const out = [];
  let acc = 0;
  for (const seg of edl) {
    const segLen = segmentDuration(seg);
    for (const w of words) {
      const overlapStart = Math.max(w.start, seg.sourceStart);
      const overlapEnd = Math.min(w.end, seg.sourceEnd);
      const overlap = overlapEnd - overlapStart;
      const wordLen = Math.max(w.end - w.start, 1e-6);
      if (overlap > 0 && overlap / wordLen >= minKeepRatio) {
        out.push({
          ...w,
          segId: seg.id,
          virtualStart: acc + (overlapStart - seg.sourceStart),
          virtualEnd: acc + (overlapEnd - seg.sourceStart),
        });
      }
    }
    acc += segLen;
  }
  return out;
}

File: backend/edl.py
"""Server-side EDL validation and rendering helpers.

An EDL is an ordered list of segments [{id, sourceStart, sourceEnd}] over the
original video's timeline. ordered_intervals() preserves list order so reordered
edits export in the arranged order.
"""
from __future__ import annotations

MIN_SEG = 0.02


def validate_edl(segments: list[dict], duration: float) -> list[dict]:
    if not isinstance(segments, list):
        raise ValueError("segments must be a list")
    for s in segments:
        try:
            a = float(s["sourceStart"])
            b = float(s["sourceEnd"])
        except (KeyError, TypeError, ValueError):
            raise ValueError(f"segment missing numeric bounds: {s!r}")
        if a < -1e-6 or b > duration + 1e-3:
            raise ValueError(f"segment {a}-{b} out of bounds [0,{duration}]")
        if b - a < MIN_SEG:
            raise ValueError(f"segment {a}-{b} shorter than {MIN_SEG}s")
    return segments


def ordered_intervals(segments: list[dict]) -> list[tuple[float, float]]:
    return [(float(s["sourceStart"]), float(s["sourceEnd"])) for s in segments]


def virtual_to_source(segments: list[dict], vt: float) -> float:
    """Map a virtual (edited-timeline) time to the original source time."""
    acc = 0.0
    for s in segments:
        a = float(s["sourceStart"])
        b = float(s["sourceEnd"])
        d = b - a
        if vt < acc + d:
            return a + (vt - acc)
        acc += d
    return float(segments[-1]["sourceEnd"]) if segments else 0.0


def cx_at(centers: list[dict], s: float) -> float:
    """Linear-interpolate the normalized face centre at source time s (0.5 = middle)."""
    prev = None
    for c in centers:
        if c.get("cx") is None:
            continue
        if c["t"] <= s:
            prev = c
        else:
            if prev is None:
                return c["cx"]
            span = c["t"] - prev["t"] or 1.0
            return prev["cx"] + ((s - prev["t"]) / span) * (c["cx"] - prev["cx"])
    return prev["cx"] if prev else 0.5


def project_words(segments: list[dict], words: list[dict], min_keep_ratio: float = 0.5) -> list[dict]:
    """Map transcript words onto the EDITED (virtual) timeline, in EDL order.

    Mirror of the frontend projectWords: walk segments in order, emit words whose
    source timestamps fall inside each segment (>=min_keep_ratio retained),
    annotated with virtual start/end. Used to time captions on the edited clip.
    """
    out: list[dict] = []
    acc = 0.0
    for seg in segments:
        a = float(seg["sourceStart"])
        b = float(seg["sourceEnd"])
        seg_len = b - a
        for w in words:
            ostart = max(w["start"], a)
            oend = min(w["end"], b)
            overlap = oend - ostart
            wlen = max(w["end"] - w["start"], 1e-6)
            if overlap > 0 and overlap / wlen >= min_keep_ratio:
                pw = {
                    "word": w["word"],
                    "virtual_start": acc + (ostart - a),
                    "virtual_end": acc + (oend - a),
                }
                if w.get("speaker") is not None:  # carry diarized speaker through
                    pw["speaker"] = w["speaker"]
                out.append(pw)
        acc += seg_len
    return out


def attach_speakers(words: list[dict], segments: list[dict]) -> list[dict]:
    """Tag each transcript word with its diarized speaker (from segment word
    ranges), so per-speaker caption colour survives projection. No-op if the
    transcript hasn't been diarized."""
    for seg in segments:
        spk = seg.get("speaker")
        if spk is None:
            continue
        for i in range(int(seg.get("word_start", 0)), int(seg.get("word_end", 0))):
            if 0 <= i < len(words):
                words[i]["speaker"] = spk
    return words
